### PR TITLE
First draft E2E tests

### PR DESCRIPTION
**Pull request by huguesBouvier**
First draft for nested E2E tests.

Check for regression, non nested E2E: https://dev.azure.com/msazure/One/_build/results?buildId=38345741&view=logs&j=4323b27c-bad1-54aa-e27b-49b281e4d21f&t=9a0ccc66-b52d-506d-9c94-6909308d8840

Nested E2E:https://dev.azure.com/msazure/One/_build/results?buildId=38379068&view=logs&j=e8afecc0-c259-5cc1-62e0-5c5d1c2a3db4&t=e8afecc0-c259-5cc1-62e0-5c5d1c2a3db4

Note: This first draft does not deploy aziot on lvl 4 + lvl 5. This will be dealt with this PR:https://github.com/Azure/iotedge/pull/4250

### CODE REVIEW

File: scripts/linux/nested-edge-deploy-agent.sh [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+    sudo sed -i "207s|.*|    image: \"${CUSTOM_EDGE_AGENT_IMAGE}\"|" /etc/iotedge/config.yaml
+
+    if [ -z $PARENT_NAME ]; then
+        sudo sed -i "208s|.*|    auth:|" /etc/iotedge/config.yaml
+        sed -i "209i\      serveraddress: \"${CONTAINER_REGISTRY}\"" /etc/iotedge/config.yaml
+        sed -i "210i\      username: \"${CONTAINER_REGISTRY_USERNAME}\"" /etc/iotedge/config.yaml
+        sed -i "211i\      password: \"${CONTAINER_REGISTRY_PASSWORD}\"" /etc/iotedge/config.yaml
+    fi
+
+    sudo cat /etc/iotedge/config.yaml
```
Comment (nyanzebra) [inline scripts/linux/nested-edge-deploy-agent.sh:54]: Does this require sudo?

File: scripts/linux/nested-edge-deploy-agent.sh [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+
+    #deploy the config in azure portal
+    az iot edge set-modules --device-id ${DEVICE_ID} --hub-name ${IOT_HUB_NAME} --content ${deployment_working_file} --output none
```
Comment (nyanzebra) [inline scripts/linux/nested-edge-deploy-agent.sh:57]: Could the output none be optional? Might be nice to see the output sometimes for debugging purposes?

File: scripts/linux/nested-edge-deploy-agent.sh [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+    echo "Start IoT edge"
+    sudo systemctl restart iotedge
+}
+
+function prepare_test_from_artifacts() {   
+    print_highlighted_message 'Prepare test from artifacts'
+
+    echo 'Remove working folder'
+    rm -rf "$working_folder"
```
Comment (nyanzebra) [inline scripts/linux/nested-edge-deploy-agent.sh:67]: Should we verify the working_folder is not root?

File: builds/e2e/templates/e2e-setup.yaml [hunk 4]
```diff
@@ -37,6 +37,9 @@ steps:
 
     Write-Output "##vso[task.setvariable variable=imageBuildId]$imageBuildId"
     Write-Output "##vso[task.setvariable variable=packageBuildId]$packageBuildId"
+
+    Write-Output '>> Package Build ID=$(az.pipeline.packages.buildId)'
+    Write-Output '>> Image Build ID=$(az.pipeline.packages.buildId)'
```
Comment (dylanbronson) [inline builds/e2e/templates/e2e-setup.yaml:42]: az.pipeline.images.buildId?

Comment (yophilav) [reply builds/e2e/templates/e2e-setup.yaml:42]: That's a pipeline variable for the build image pipeline. ( a custom defined var)

File: scripts/linux/nested-edge-deploy-agent.sh (hunk 1 above)
Comment (huguesBouvier) [reply scripts/linux/nested-edge-deploy-agent.sh:54]: Yes, files in /etc requires sudo access for reading

File: builds/e2e/nested-e2e.yaml [hunk 5]
```diff
@@ -10,31 +10,111 @@ resources:
...
+      - template: templates/lock-test-agent.yaml     
+      - template: templates/e2e-clean-directory.yaml
+      - template: templates/nested-get-root-ca.yaml
+      - template: templates/nested-get-device-name.yaml    
+      - template: templates/e2e-setup.yaml
+      - template: templates/e2e-clear-docker-cached-images.yaml
+      - template: templates/nested-e2e-run.yaml
+  - job:  Clean_up     
+    dependsOn: SetupVM_and_RunTest_level3
+    condition: true
```
Comment (yophilav) [inline builds/e2e/nested-e2e.yaml:101]: `condition: always()`

File: scripts/linux/nested-edge-deploy-agent.sh (hunk 2 above)
Comment (huguesBouvier) [reply scripts/linux/nested-edge-deploy-agent.sh:57]: I enabled it but it doesn't ring much to the table. When there is a failure you get display anyway

File: scripts/linux/nested-edge-deploy-agent.sh (hunk 3 above)
Comment (huguesBouvier) [reply scripts/linux/nested-edge-deploy-agent.sh:67]: It is an automated test, that folder is not user selected. You don't need to make sanity check from the inside. Sanity checks are good practice for external interactions

File: builds/e2e/templates/nested-create-identity.yaml [hunk 6]
```diff
@@ -0,0 +1,41 @@
+steps: 
+  - task: Bash@3
+    displayName: 'Az login'
+    inputs:
+      targetType: inline
+      script: |     
+        az login --service-principal -p $(edgebuild-service-principal-secret) -u $(servicePrincipal.clientId) --tenant $(servicePrincipal.tenantId)
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-create-identity.yaml:7]: Pipeline will not print secret. Have a look in the logs

File: builds/e2e/nested-e2e.yaml [hunk 7]
```diff
@@ -10,31 +10,111 @@ resources:
...
+  jobs:
+  - job: SetupVM_level5
+    displayName: SettingUp level 5
+    condition: ne(variables['agent.group'], '')
+    timeoutInMinutes: 180
+    variables:
+     artifactName: iotedged-ubuntu16.04-amd64
+     vsts.project: $(System.TeamProjectId)
+     level: '5'
+     deploymentFile: 'nestededge_topLayerBaseDeployment.json'
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/nested-e2e.yaml:24]: That would not the case for the other yaml file before nested E2E tests. We should first have eveything consistant with prior strategy then discuss improvements.

File: builds/e2e/templates/nested-get-secrets.yaml [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+    inputs:
+      azureSubscription: $(az.subscription)
+      KeyVaultName: $(azure.keyVault)
+      SecretsFilter: 'edgebuilds-azurecr-io-username,edgebuilds-azurecr-io-pwd,kvLogAnalyticWorkspaceId,kvLogAnalyticSharedKey,EdgeConnectivityStorageAccountConnString'
+  - task: AzureKeyVault@1
+    displayName: 'Nested Edge Azure Key Vault'
+    inputs:
+      azureSubscription: $(az.subscription)
+      KeyVaultName: $(azure.keyVault)
+      SecretsFilter: 'ConnectivityNestedEdge-ConnectionString,ConnectivityNestedEdge-EventHubConnectionString,edgebuild-blob-core-connection-string,edgebuild-service-principal-secret,edgebuild-1-PAT'
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-get-secrets.yaml:13]: I am not sure, I think it is better to have them all at the same place. If we start to have more config than nested, we don't want a new keyvault for each one.

File: builds/e2e/templates/nested-get-root-ca.yaml [hunk 9]
```diff
@@ -0,0 +1,12 @@
+steps:
+  - task: Bash@3
+    displayName: 'Get rootCA'
+    inputs:
+      targetType: inline
+      script: |
+        az storage blob download --file rootCA.tar.bz2 --container-name test-certificates --name test-certs.tar.bz2 --connection-string "$(edgebuild-blob-core-connection-string)"
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-get-root-ca.yaml:7]: Will do

File: builds/e2e/templates/nested-get-secrets.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+steps:
+  - task: AzureKeyVault@1
+    displayName: 'EdgeBuild Azure Key Vault'
+    inputs:
+      azureSubscription: $(az.subscription)
+      KeyVaultName: $(azure.keyVault)
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-get-secrets.yaml:6]: Will do

File: builds/e2e/templates/nested-get-root-ca.yaml (hunk 9 above)
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-get-root-ca.yaml:7]: https://msazure.visualstudio.com/One/_workitems/edit/9142565

File: scripts/linux/nested-edge-deploy-agent.sh [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+connectivity_deployment_artifact_file="e2e_deployment_files/$DEPLOYMENT_FILE_NAME"
+deployment_working_file="$working_folder/deployment.json"
+
+test_setup
+create_certificates
+install_and_setup_iotedge
+set_output_params
+
+#clean up
+#az iot hub device-identity delete -n ${iotHubName} -d ${iotEdgeDevicesName}
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply scripts/linux/nested-edge-deploy-agent.sh:254]: Fixed

File: builds/e2e/nested-e2e.yaml (hunk 5 above)
Comment (huguesBouvier) [reply builds/e2e/nested-e2e.yaml:101]: Fixed

File: builds/e2e/nested-e2e.yaml [hunk 12]
```diff
@@ -10,31 +10,111 @@ resources:
...
 
+  - job:  SetupVM_and_RunTest_level3
+    dependsOn: SetupVM_level4
+    displayName: Set up and run tests
+    timeoutInMinutes: 180
     variables:
+      parentName: $[ dependencies.SetupVM_level4.outputs['deployIoTEdge.deviceName'] ]
+      parentDeviceId: $[ dependencies.SetupVM_level4.outputs['createIdentity.parentDeviceId'] ] 
+#      parentName: nestededgetest02.eastus2.cloudapp.azure.com
+#      parentDeviceId: level_4_20201210.99
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/nested-e2e.yaml:77]: Fixed

File: builds/e2e/nested-e2e.yaml [hunk 13]
```diff
@@ -10,31 +10,111 @@ resources:
...
+       - Agent.OS -equals Linux
+       - Agent.OSArchitecture -equals X64
+       - run-idle -equals true
+       - status -equals unlocked
+    steps:
+     - template: templates/nested-get-secrets.yaml
+     - template: templates/lock-test-agent.yaml
+     - template: templates/nested-get-root-ca.yaml       
+     - template: templates/nested-create-identity.yaml   
+     - template: templates/nested-agent-deploy.yaml
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/nested-e2e.yaml:40]: Cleaning already in nested-edge-deploy-agent

File: builds/e2e/templates/nested-create-identity.yaml [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+            echo "Attaching device Id to parent device Id: $(parentDeviceId)"
+
+            az iot hub device-identity create -n ${iotHubName} -d ${deviceId} --ee --pd $(parentDeviceId) --output none
+        fi
+        connectionString=$(az iot hub device-identity connection-string show -d ${deviceId} -n ${iotHubName} --query 'connectionString' -o tsv)
+
+        #set output variable to use later in the job
+        echo "##vso[task.setvariable variable=deviceId]${deviceId}"
+        echo "##vso[task.setvariable variable=iotHubName]${iotHubName}"
+        echo "##vso[task.setvariable variable=connectionString]${connectionString}"
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-create-identity.yaml:37]: Yes, I realized if you put isOutput=true it doesn't work locally. If you don't put isOutput=true it doesn't work between job. If you want both you have to add both

File: builds/e2e/templates/nested-agent-deploy.yaml [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+      TargetFolder: "$(Agent.HomeDirectory)/../artifacts/$(az.pipeline.images.artifacts)"
+      CleanTargetFolder: true 
+  - task: Bash@3
+    displayName: 'Deployment iotedge on agent'
+    name: deployIoTEdge
+    inputs:
+      targetType: inline
+      #@TODO change Build.Repository.LocalPath to use artifact.
+      #@TODO remove hard coding of deploymentFileName: It is also hardcoded in install.sh
+      #@TODO remove hard coding of level and parent
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply builds/e2e/templates/nested-agent-deploy.yaml:68]: yes. Removed

File: scripts/linux/nested-edge-deploy-agent.sh [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+
+function create_certificates() {
+    echo "Installing test root certificate bundle."    
+
+    echo "Generating edge device certificate"
+    device_name=$(az vm show -d  -g "iotedge-deploy" -n $(hostname) --query fqdns)
+    eval device_name=${device_name}
+    echo "  Hostname FQDN: ${device_name}" 
+
+    /certs/certGen.sh create_edge_device_certificate ${device_name}
```
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply scripts/linux/nested-edge-deploy-agent.sh:13]: They don't because they have the same name. But agreed. Will fix that.

**Review by varunpuranik — APPROVED** [review]

File: scripts/linux/nested-edge-deploy-agent.sh (hunk 16 above)
[the remark this answers is not in the document]
Comment (huguesBouvier) [reply scripts/linux/nested-edge-deploy-agent.sh:13]: Certificate are cleaned

**Pull request merged**